### PR TITLE
Add Time Service

### PR DESCRIPTION
**Pull request by Akoszewski**

### CODE REVIEW

File: src/TimeService/timeservice-api.c [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#include "timeservice-api.h"
+
+static uint64_t starting_time;
+
+void nanoseconds_to_cuc_timestamp(uint64_t nanoseconds, CUCTimestamp *timestamp)
+{
+   long double seconds = (long double)nanoseconds / NS_PER_SECOND;
+   uint64_t seconds_shifted = seconds * 256 * 256 * 256;
```
Comment (Ficzer) [inline src/TimeService/timeservice-api.c:8]: Remove magic numbers

File: src/TimeService/timeservice-api.c [hunk 2]
```diff
@@ -0,0 +1,85 @@
+#include "timeservice-api.h"
+
+static uint64_t starting_time;
+
+void nanoseconds_to_cuc_timestamp(uint64_t nanoseconds, CUCTimestamp *timestamp)
```
Comment (Ficzer) [inline src/TimeService/timeservice-api.c:5]: Add TimeService context to make code pseudo objective

File: src/TimeService/timeservice-api.c [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+    }
+}
+
+
+void TimeService_ClockStatus(ClockStatusEnum *status)
+{
+   *status = STOPPED;
+
+   uint64_t start_time = get_clock();
+   delay(10000);
```
Comment (Ficzer) [inline src/TimeService/timeservice-api.c:60]: is waiting 10 cycles good idea to check clock status?

File: src/TimeService/timeservice-api.h [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+#endif
+
+#ifdef GENERIC_LINUX_TARGET
+#include "timeservice-linux.h"
+#endif
+
+void TimeService_Startup(void);
+
+
+void TimeService_CfsTimestampCmp(const CfsTimestamp *t1, const CfsTimestamp *t2, int *result);
```
Comment (Ficzer) [inline src/TimeService/timeservice-api.h:22]: can't we just return the value result, do we have to have void function everywhere?

File: src/TimeService/timeservice-rtems.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+   Timer_Apbctrl1_init(timerId, timer, defaultInterruptHandler);
+   Timer_Apbctrl1_setBaseScalerReloadValue(timer, SCALER_VALUE);
+   Timer_Apbctrl1_setConfigRegisters(timer, &config);
+   Timer_Apbctrl1_start(timer);
+}
+
+static uint64_t swap_bytes(uint64_t x)
+{
+  uint64_t result;
+  result = ((x & 0x00000000000000FFULL) << 56) |
```
Comment (Ficzer) [inline src/TimeService/timeservice-rtems.h:24]: At least small comment what is happening here would be nice

File: src/TimeService/timeservice-rtems.h [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+   long double ticks_per_nanosecond = (long double)TICKS_PER_SECOND / NS_PER_SECOND;
+   uint64_t nanoseconds = ticks / ticks_per_nanosecond;
+   return nanoseconds;
+}
+
+#define INIT_TIMERS() \
+    init_timer(&timer_2, true, Timer_Id_3); \
+    init_timer(&timer_1, false, Timer_Id_2)
+
+#define ADJUST_ENDIANESS(x) \
```
Comment (Ficzer) [inline src/TimeService/timeservice-rtems.h:48]: do we need this macro?

File: src/TimeService/timeservice-utils.h [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+   uint64_t subseconds;
+} CfsTimestamp;
+
+typedef enum {
+    UNAVAILABLE = 0,
+    STOPPED = 1,
+    STARTED = 2
+} ClockStatusEnum;
+
+static int cmp_memory(const unsigned char *str1, const unsigned char *str2, int len)
```
Comment (Ficzer) [inline src/TimeService/timeservice-utils.h:29]: I would try to use full names of an arguments where it is possible (for example "length" not "len")

File: src/TimeService/timeservice-rtems.h [hunk 8]
```diff
@@ -0,0 +1,49 @@
+#include "interfaces_info.h"
+
+#define TICKS_PER_SECOND 1000000
+#define SCALER_VALUE 5
+
+void init_timer(Timer_Apbctrl1 *const timer, bool isChained, Timer_Id timerId)
```
Comment (Ficzer) [inline src/TimeService/timeservice-rtems.h:6]: Do we need to restart timers at all? We have two timers running in chained configuration right now in kazoo, they start always when starting the program. Restarting them may lead to problems in my opinion. Moreover i think it is unnecessary if as far as i know in TimeService we only measure the passage of time.

File: src/TimeService/timeservice-linux.h [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+
+#include <time.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <string.h>
+
+#define INIT_TIMERS()
+#define ADJUST_ENDIANESS(x)
+
+uint64_t get_clock()
```
Comment (Ficzer) [inline src/TimeService/timeservice-linux.h:11]: We are returning nanoseconds but function name is "get_clock", shouldn't it be "get_time" or something similar?

File: src/TimeService/timeservice-rtems.h [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+uint64_t get_clock()
+{
+   uint64_t ticks = UINT64_MAX - ((uint64_t)(Timer_Apbctrl1_getCounterValue(&timer_2) + 1) << 32
+         | Timer_Apbctrl1_getCounterValue(&timer_1));
+   long double ticks_per_nanosecond = (long double)TICKS_PER_SECOND / NS_PER_SECOND;
+   uint64_t nanoseconds = ticks / ticks_per_nanosecond;
+   return nanoseconds;
+}
+
+#define INIT_TIMERS() \
```
Comment (Ficzer) [inline src/TimeService/timeservice-rtems.h:44]: Do we need this macro?

**Review by Ficzer — COMMENTED** [review]
Remove magic numbers

File: src/TimeService/timeservice-rtems.h (hunk 5 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-rtems.h:24]: It's just swapping endianess

File: src/TimeService/timeservice-api.h (hunk 4 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-api.h:22]: We can it's just architecture based on the interfaces in taste but in the API I can change it

File: src/TimeService/timeservice-rtems.h (hunk 8 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-rtems.h:6]: In the middle of the development I noticed that the clock is returning 0 and restarting timers helped but now I see that it somehow works without that, so I will remove it.

File: src/TimeService/timeservice-rtems.h (hunk 10 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-rtems.h:44]: I will delete it like I told under the comment about restarting timers

File: src/TimeService/timeservice-rtems.h (hunk 6 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-rtems.h:48]: Yes, it's useful when a code requires swapping bytes only on some architectures

File: src/TimeService/timeservice-utils.h (hunk 7 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-utils.h:29]: OK

File: src/TimeService/timeservice-linux.h (hunk 9 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-linux.h:11]: OK

File: src/TimeService/timeservice-api.c (hunk 1 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-api.c]: Done

File: src/TimeService/timeservice-api.c (hunk 2 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-api.c:5]: Done

File: src/TimeService/timeservice-rtems.h (hunk 5 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-rtems.h:13]: I added a comment

File: src/TimeService/timeservice-rtems.h (hunk 6 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-rtems.h]: I used a function instead

File: src/TimeService/timeservice-api.h (hunk 4 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-api.h]: I changed the convention wherever it was possible

File: src/TimeService/timeservice-api.c (hunk 3 above)
Comment (Akoszewski) [reply src/TimeService/timeservice-api.c]: I will try to use clock_nanosleep and rtems version of it but the whole function is not well described in D5 and I am not sure what that function should actually do and return

Comment (Ficzer) [reply src/TimeService/timeservice-api.c]: I still dont understand why we check clock status this way. Using delay to check if clock is running seems a big anti-pattern to me. Is not there any flag or can't we introduce one to check for clock status?

**Review by Ficzer — APPROVED** [review]

**Pull request closed without merging**